src/mz/src/bin/mz/upgrader.rs

```diff
@@ -0,0 +1,289 @@
+        reader.read_line(&mut reg_version)?;
+
+        let stored_timestamp: u64 = timestamp
+            .trim()
+            .parse()
+            .map_err(|_| Error::ParsingTimestampU64Error)?;
+
+        Ok(Some((stored_timestamp, reg_version)))
+    }
+
+    /// Returns true if the installed version of `mz`
+    /// is older than the version sent by parameter.
+    fn is_installed_version_older_than(&self, (major, minor, patch): (u64, u64, u64)) -> bool {
+        let local_version = build_info!().semver_version();
+
+        if major > local_version.major {
+            return true;
+        }
+
+        if minor > local_version.minor {
+            return true;
+        }
+
+        if major == local_version.major
+            && minor == local_version.minor
+            && patch > local_version.patch + 5
+        {
+            return true;
+        }
+
+        false
+    }
+
+    /// Fetches and returns the latest tag version from the Materialize repository
+    /// using the GitHub public API.
+    async fn get_last_tag_version(&self) -> Result<Option<(u64, u64, u64)>, Error> {
+        #[derive(Deserialize, Debug)]
+        struct GithubTag {
+            name: String,
+        }
+
+        // Headers recommended by GitHub: https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repository-tags
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            ACCEPT,
+            HeaderValue::from_static("application/vnd.github+json"),
+        );
+        // The GitHub API endpoint requires setting the user agent; otherwise, it returns a 403.
+        let user_agent = format!("mz/{}", VERSION.clone());
+        headers.insert(
+            USER_AGENT,
+            HeaderValue::from_str(&user_agent).map_err(Error::HeaderParseError)?,
+        );
+
+        let client = reqwest::Client::new();
+
+        // As of the creation date of this code, the last `mz-vx.y.z` tag is located on page 4 (last-one).
+        // I kept the first page in the event to avoid any issues.
+        let mut page = 1;
+
+        let last_tag = loop {
+            let url = format!(
+                "https://api.github.com/repos/materializeInc/materialize/tags?per_page=100&page={}",
+                page
+            );
+            let response = client
+                .get(url)
+                .headers(headers.clone())
+                .send()
+                .await
+                .map_err(Error::GitHubFetchError)?;
+
+            let tags: Vec<GithubTag> = response
+                .json()
+                .await
+                .map_err(Error::ReqwestJsonParseError)?;
+
+            if let Some(tag) = tags
+                .into_iter()
+                .find(|tag| tag.name.starts_with(MZ_V_PREFIX))
+            {
+                break tag;
+            }
+            page += 1;
+        };
+
+        let last_version = self.strip_version(last_tag.name)?;
+
+        Ok(last_version)
```

Comment:
Ah, while this will work, I have an easier method to suggest! Let's just write the latest `mz` version to a file on S3, so you can `curl https://binaries.materialize.com/mz-latest.version` and get that version directly. We already _basically_ do this when we upload the "latest" redirect that always points at the latest binary version of `mz`:

```
$ curl -I https://binaries.materialize.com/mz-latest-x86_64-unknown-linux-gnu.tar.gz
HTTP/2 301 
content-length: 0
date: Thu, 31 Aug 2023 03:37:38 GMT
location: /mz-v0.1.3-x86_64-unknown-linux-gnu.tar.gz
server: AmazonS3
x-cache: Miss from cloudfront
via: 1.1 b787d00b53c710066f0d489ebc435a74.cloudfront.net (CloudFront)
x-amz-cf-pop: BOS50-P3
x-amz-cf-id: FBxAqGsCErVnG6860alfIA-uCZ6qnp81BN0Sqsr7K1qhNO4KZqD7bg==
```

Here's where we upload that file today:

https://github.com/MaterializeInc/materialize/blob/7ee8a2891c6a52eecbcf281cddb408b75a263a03/ci/deploy_mz/linux.py#L48

Should be easy to add some logic there that uploads a file containing _just_ the version to the key `mz-latest.version`.

Reply:
We could also just `curl mz-latest` as above, and read the `Location` header?

Reply:
@necaris this works! To make sure we do not break this logic in the future I will comment the fn `upload_latest_redirect()` https://github.com/MaterializeInc/materialize/blob/ec72b724dd5af26ece8763273c813b907b8aba27/ci/deploy_mz/deploy_util.py#L70

Reply:
> We could also just `curl mz-latest` as above, and read the `Location` header?

Ah, sorry, I don't think that's a good approach, which is why I suggested adding a new file dedicated to this purpose. 

This kind of code is quite brittle: 

https://github.com/MaterializeInc/materialize/pull/21459/files#diff-43c0df1ed442e7551486108c76d553ed34a9a550b78d0b7c28b0c1144bcb21a4R112-R116

If the naming scheme for the binaries changes, this is liable to go off the rails.

Even with a comment in deploy_util.py, we're still coupling two fairly unrelated things together (getting the latest version for any architecture, vs downloading the latest binary for a given architecture). We control the full deploy process of `mz`, and the code is easy to change, so we should just change it to exactly suit our needs. When the full system is under our control and easy to change, the best solution is the one that minimizes the net complexity of the system as a whole.

Reply:
I put up https://github.com/MaterializeInc/materialize/pull/21563 to create said file during `mz` releases.